flink-connector-rabbitmq/src/main/java/org/apache/flink/connector/rabbitmq/sink/RabbitMQSinkWriter.java

```diff
@@ -0,0 +1,260 @@
+    @Override
+    public void write(T t, Context context) throws IOException, InterruptedException {
+        awaitInflightMessagesBelow(maximumInflightMessages);
+        RabbitMQMessage<T> recordMessage = messageConverter.toRabbitMQMessage(t);
+        publishMessage(recordMessage);
+    }
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException, InterruptedException {
+        awaitInflightMessagesBelow(1);
+    }
+
+    @Override
+    public void close() throws Exception {
+        awaitInflightMessagesBelow(1);
+        channel.close();
+        connection.close();
+    }
+
+    private void awaitInflightMessagesBelow(int maximumInflightMessages)
+            throws InterruptedException {
+        while (inflightMessages.size() >= maximumInflightMessages) {
+            mailboxExecutor.yield();
+        }
+    }
+
+    private void publishMessage(RabbitMQMessage<T> recordMessage) throws IOException {
+        byte[] message = serializationSchema.serialize(recordMessage.getMessage());
+        Long sequenceNumber = channel.getNextPublishSeqNo();
+
+        if (messageConverter.supportsExchangeRouting()) {
+            Preconditions.checkArgument(
+                    returnListener != null
+                            || !(recordMessage.isImmediate() || recordMessage.isMandatory()),
+                    "Return listener must be set if immediate or mandatory delivery is requested");
+
+            channel.basicPublish(
+                    recordMessage.getExchange(),
+                    recordMessage.getRoutingKey(),
+                    recordMessage.isMandatory(),
+                    recordMessage.isImmediate(),
+                    recordMessage.getMessageProperties(),
+                    message);
+
+        } else {
+            channel.basicPublish(
+                    recordMessage.getExchange(),
+                    queueName,
+                    recordMessage.getMessageProperties(),
+                    message);
+        }
+
+        inflightMessages.put(sequenceNumber, recordMessage);
+    }
+
+    private ConfirmCallback getSuccessConfirmCallback() {
+        return (seqNo, acknowledgePrevious) ->
+                mailboxExecutor.execute(
+                        () -> {
+                            if (acknowledgePrevious) {
+                                LOG.debug(
+                                        "Acknowledge all messages with sequence number less than or equal to {}",
+                                        seqNo);
+                                while (!inflightMessages.isEmpty()
+                                        && inflightMessages.firstKey() <= (Long) seqNo) {
+                                    RabbitMQMessage<T> message =
+                                            inflightMessages.remove(inflightMessages.firstKey());
+                                    numBytesOutCounter.inc(
+                                            serializationSchema.serialize(message.getMessage())
```

Comment:
I would suggest to have another record for messages with size. We already serialize the message on publish, so that we can avoid serialization again here when updating counters